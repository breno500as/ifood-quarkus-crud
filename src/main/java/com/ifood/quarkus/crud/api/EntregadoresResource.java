package com.ifood.quarkus.crud.api;

import org.eclipse.microprofile.openapi.annotations.tags.Tag;

import com.ifood.quarkus.crud.entity.Entregador;

import io.quarkus.hibernate.orm.rest.data.panache.PanacheEntityResource;

@Tag(name = "entregadores")
public interface EntregadoresResource extends PanacheEntityResource<Entregador, Long> {

}
