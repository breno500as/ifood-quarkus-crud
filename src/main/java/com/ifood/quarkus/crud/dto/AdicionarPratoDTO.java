package com.ifood.quarkus.crud.dto;

import java.math.BigDecimal;

import javax.validation.constraints.NotNull;


public class AdicionarPratoDTO {
	

	@NotNull
	public String nome;

	@NotNull
	public String descricao;

	public BigDecimal preco;

}
