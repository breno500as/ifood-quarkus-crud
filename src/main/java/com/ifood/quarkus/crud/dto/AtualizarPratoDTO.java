package com.ifood.quarkus.crud.dto;

import java.math.BigDecimal;

public class AtualizarPratoDTO {

	   public BigDecimal preco;
}
