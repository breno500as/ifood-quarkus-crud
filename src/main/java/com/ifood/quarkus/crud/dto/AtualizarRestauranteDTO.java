package com.ifood.quarkus.crud.dto;

import javax.validation.constraints.NotNull;

public class AtualizarRestauranteDTO {
	
	@NotNull
	public String nomeFantasia;

}
