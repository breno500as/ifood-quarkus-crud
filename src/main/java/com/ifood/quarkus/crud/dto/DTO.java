package com.ifood.quarkus.crud.dto;

import javax.validation.ConstraintValidatorContext;

public interface DTO {

	default boolean isValid(ConstraintValidatorContext v) {
		return true;
	}
}
