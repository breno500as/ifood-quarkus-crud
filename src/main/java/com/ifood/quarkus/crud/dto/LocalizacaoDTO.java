package com.ifood.quarkus.crud.dto;

public class LocalizacaoDTO {

	public Double latitude;

	public Double longitude;

}
