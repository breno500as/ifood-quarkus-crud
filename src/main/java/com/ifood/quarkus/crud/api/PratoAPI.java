package com.ifood.quarkus.crud.api;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import javax.inject.Inject;
import javax.transaction.Transactional;
import javax.validation.Valid;
import javax.ws.rs.Consumes;
import javax.ws.rs.DELETE;
import javax.ws.rs.GET;
import javax.ws.rs.NotFoundException;
import javax.ws.rs.POST;
import javax.ws.rs.PUT;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.Response.Status;

import org.eclipse.microprofile.metrics.annotation.Counted;
import org.eclipse.microprofile.metrics.annotation.SimplyTimed;
import org.eclipse.microprofile.metrics.annotation.Timed;
import org.eclipse.microprofile.openapi.annotations.tags.Tag;

import com.ifood.quarkus.crud.dto.AdicionarPratoDTO;
import com.ifood.quarkus.crud.dto.AtualizarPratoDTO;
import com.ifood.quarkus.crud.dto.PratoDTO;
import com.ifood.quarkus.crud.dto.PratoMapper;
import com.ifood.quarkus.crud.entity.Prato;
import com.ifood.quarkus.crud.entity.Restaurante;

@Path("restaurantes/{idRestaurante}/pratos")
@Tag(name = "pratos")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
public class PratoAPI {

	@Inject
	PratoMapper pratoMapper;

	@GET
	@Counted(name = "Número de chamadas buscarPratos") // Conta quantas vezes o método foi chamado
	@SimplyTimed(name = "Tempo simples para executar buscarPratos")// Mostra o tempo simples do método
	@Timed(name = "Tempo total para executar buscarPratos")// Mostra o tempo total do método
	//@Gauge Pega informações de recursos, quantidade de memória, espaço em disco, temperatura do processador, número de elementos em uma fila, número de elementos em uma tabela
	public List<PratoDTO> buscarPratos(@PathParam("idRestaurante") Long idRestaurante) {
		Optional<Restaurante> restauranteOp = Restaurante.findByIdOptional(idRestaurante);
		if (restauranteOp.isEmpty()) {
			throw new NotFoundException("Restaurante não existe");
		}
		Stream<Prato> pratos = Prato.stream("restaurante", restauranteOp.get());
		return pratos.map(p -> pratoMapper.toDTO(p)).collect(Collectors.toList());
	}

	@POST
	@Transactional
	public Response adicionarPrato(@PathParam("idRestaurante") Long idRestaurante, @Valid AdicionarPratoDTO dto) {
		Optional<Restaurante> restauranteOp = Restaurante.findByIdOptional(idRestaurante);
		if (restauranteOp.isEmpty()) {
			throw new NotFoundException("Restaurante não existe");
		}
		// //Utilizando dto, recebi detached entity passed to persist:
		// Prato prato = new Prato();
		// prato.nome = dto.nome;
		// prato.descricao = dto.descricao;
		//
		// prato.preco = dto.preco;
		// prato.restaurante = restauranteOp.get();
		// prato.persist();

		Prato prato = pratoMapper.toPrato(dto);
		prato.restaurante = restauranteOp.get();
		prato.persist();
		return Response.status(Status.CREATED).build();
	}

	@PUT
	@Path("{id}")
	@Transactional
	public void atualizar(@PathParam("idRestaurante") Long idRestaurante, @PathParam("id") Long id,
			@Valid AtualizarPratoDTO dto) {
		Optional<Restaurante> restauranteOp = Restaurante.findByIdOptional(idRestaurante);
		if (restauranteOp.isEmpty()) {
			throw new NotFoundException("Restaurante não existe");
		}

		// No nosso caso, id do prato vai ser único, independente do restaurante...
		Optional<Prato> pratoOp = Prato.findByIdOptional(id);

		if (pratoOp.isEmpty()) {
			throw new NotFoundException("Prato não existe");
		}
		Prato prato = pratoOp.get();
	    pratoMapper.toPrato(dto, prato);
		prato.persist();
	}

	@DELETE
	@Path("{id}")
	@Transactional
	public void delete(@PathParam("idRestaurante") Long idRestaurante, @PathParam("id") Long id) {
		Optional<Restaurante> restauranteOp = Restaurante.findByIdOptional(idRestaurante);
		if (restauranteOp.isEmpty()) {
			throw new NotFoundException("Restaurante não existe");
		}

		Optional<Prato> pratoOp = Prato.findByIdOptional(id);

		pratoOp.ifPresentOrElse(Prato::delete, () -> {
			throw new NotFoundException();
		});
	}

}
