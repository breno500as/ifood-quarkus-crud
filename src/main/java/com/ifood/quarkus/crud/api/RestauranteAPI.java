package com.ifood.quarkus.crud.api;

import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import javax.annotation.security.RolesAllowed;
import javax.inject.Inject;
import javax.transaction.Transactional;
import javax.validation.Valid;
import javax.ws.rs.Consumes;
import javax.ws.rs.DELETE;
import javax.ws.rs.ForbiddenException;
import javax.ws.rs.GET;
import javax.ws.rs.NotFoundException;
import javax.ws.rs.POST;
import javax.ws.rs.PUT;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.Response.Status;

import org.eclipse.microprofile.jwt.Claim;
import org.eclipse.microprofile.jwt.Claims;
import org.eclipse.microprofile.jwt.JsonWebToken;
import org.eclipse.microprofile.openapi.annotations.enums.SecuritySchemeType;
import org.eclipse.microprofile.openapi.annotations.media.Content;
import org.eclipse.microprofile.openapi.annotations.media.Schema;
import org.eclipse.microprofile.openapi.annotations.responses.APIResponse;
import org.eclipse.microprofile.openapi.annotations.security.OAuthFlow;
import org.eclipse.microprofile.openapi.annotations.security.OAuthFlows;
import org.eclipse.microprofile.openapi.annotations.security.SecurityRequirement;
import org.eclipse.microprofile.openapi.annotations.security.SecurityRequirements;
import org.eclipse.microprofile.openapi.annotations.security.SecurityScheme;
import org.eclipse.microprofile.openapi.annotations.tags.Tag;
import org.eclipse.microprofile.reactive.messaging.Channel;
import org.eclipse.microprofile.reactive.messaging.Emitter;

import com.ifood.quarkus.crud.dto.AdicionarRestauranteDTO;
import com.ifood.quarkus.crud.dto.AtualizarRestauranteDTO;
import com.ifood.quarkus.crud.dto.RestauranteDTO;
import com.ifood.quarkus.crud.dto.RestauranteMapper;
import com.ifood.quarkus.crud.entity.Restaurante;
import com.ifood.quarkus.crud.exception.ConstraintViolationResponse;

@Path("/restaurantes")
@Produces(MediaType.APPLICATION_JSON)
@Consumes(MediaType.APPLICATION_JSON)
@Tag(name = "restaurantes")
@RolesAllowed("proprietario")
@SecurityScheme(securitySchemeName = "ifood-oauth", type = SecuritySchemeType.OAUTH2, flows = @OAuthFlows(password = @OAuthFlow(tokenUrl = "http://localhost:8180/auth/realms/ifood/protocol/openid-connect/token")))
@SecurityRequirements(value = { @SecurityRequirement(name = "ifood-oauth", scopes = {}) })
public class RestauranteAPI {

	@Inject
	private RestauranteMapper restauranteMapper;

	@Inject
	@Channel("restaurantes")
	private Emitter<Restaurante> emitter;
	
	@Inject
	private JsonWebToken token; // Objeto que representa o conteúdo do token (header / payload)

	@Claim(value = "preferred_username")
	private String nomeProprietarioToken; // String que representa o claim preferred_username do token (nome do usuário logado do token)
	
	@Claim(standard = Claims.sub)
	private String sub;  // String que representa o claim subject do token (subject do usuário logado do token)
	
	@GET
	public List<RestauranteDTO> listAll() {
		Stream<Restaurante> restaurantes = Restaurante.streamAll();
		return restaurantes.map(r -> restauranteMapper.toRestauranteDTO(r)).collect(Collectors.toList());
	}

	@POST
	@Transactional
	@APIResponse(responseCode = "201", description = "Caso restaurante seja cadastrado com sucesso")
	@APIResponse(responseCode = "400", content = @Content(schema = @Schema(allOf = ConstraintViolationResponse.class)))
	public Response cria(@Valid AdicionarRestauranteDTO restauranteDTO) {
		Restaurante r = this.restauranteMapper.toRestaurante(restauranteDTO);
		r.proprietario = this.nomeProprietarioToken;
		r.persist();
		this.emitter.send(r);
		return Response.status(Status.CREATED).build();
	}

	@PUT
	@Path("{id}")
	@Transactional
	public Response atualiza(@PathParam("id") Long id, @Valid AtualizarRestauranteDTO dto) {
		Optional<Restaurante> restauranteOp = Restaurante.findByIdOptional(id);
		if (restauranteOp.isEmpty()) {
			throw new NotFoundException();
		}
		Restaurante restaurante = restauranteOp.get();

		 if (!restaurante.proprietario.equals(this.nomeProprietarioToken)) {
		 throw new ForbiddenException();
		 }

		// MapStruct: aqui passo a referencia para ser atualizada
		this.restauranteMapper.toRestaurante(dto, restaurante);

		restaurante.persist();
		
		return Response.ok().build();
	}

	@DELETE
	@Path("{id}")
	@Transactional
	public Response delete(@PathParam("id") Long id) {
		Optional<Restaurante> optional = Restaurante.findByIdOptional(id);

		optional.ifPresentOrElse(Restaurante::delete, () -> {
			throw new NotFoundException();
		});

		return Response.noContent().build();
	}

}
