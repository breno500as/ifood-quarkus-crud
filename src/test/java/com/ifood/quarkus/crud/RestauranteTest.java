package com.ifood.quarkus.crud;

import javax.ws.rs.core.Response.Status;

import org.approvaltests.Approvals;
import org.junit.Assert;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import com.github.database.rider.cdi.api.DBRider;
import com.github.database.rider.core.api.configuration.DBUnit;
import com.github.database.rider.core.api.configuration.Orthography;
import com.github.database.rider.core.api.dataset.DataSet;
import com.ifood.quarkus.crud.dto.AtualizarRestauranteDTO;
import com.ifood.quarkus.crud.entity.Restaurante;

import io.quarkus.test.common.QuarkusTestResource;
import io.quarkus.test.junit.QuarkusTest;
import io.restassured.RestAssured;
import io.restassured.http.ContentType;
import io.restassured.http.Header;
import io.restassured.specification.RequestSpecification;

@QuarkusTest
@DBRider
@DBUnit(caseSensitiveTableNames = false,caseInsensitiveStrategy = Orthography.LOWERCASE)
@QuarkusTestResource(DatabaseLifecycle.class)
public class RestauranteTest {
	
	private String token;

	@Test
	@DataSet(value = "restaurantes1.yml")
	public void getRestaurantes() {
		String resultado = given().when().get("/restaurantes").then().statusCode(200).extract().asString();
		Approvals.verifyJson(resultado);
	}
	
	@BeforeEach
	public void getToken() throws Exception {
		this.token = TokenUtils.generateTokenString("/JWTProprietarioClaims.json", null);
	}

	private RequestSpecification given() {
		return RestAssured.given().contentType(ContentType.JSON).header(new Header("Authorization", "Bearer " + token));
	}

	// @Test
	// @DataSet("restaurantes1.yml")
	public void testAlterarRestaurante() {
		AtualizarRestauranteDTO dto = new AtualizarRestauranteDTO();
		dto.nome = "novoNome";
		Long parameterValue = 123L;
		
		given().with().pathParam("id", parameterValue).body(dto).when().put("/restaurantes/{id}").then()
				.statusCode(Status.NO_CONTENT.getStatusCode()).extract().asString();

		Restaurante findById = Restaurante.findById(parameterValue);

		// poderia testar todos os outros atribudos
		Assert.assertEquals(dto.nome, findById.nome);

	}
}
