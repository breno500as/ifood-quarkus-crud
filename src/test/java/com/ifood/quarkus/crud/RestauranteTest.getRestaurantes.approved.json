[
  {
    "id": 1,
    "nome": "bertos"
  }
]